lib/vnet/setup.go

```diff
@@ -31,101 +30,142 @@ import (
 	"github.com/gravitational/teleport/api/types"
 )
 
-// Setup spins up a separate process with the admin subcommand to create a TUN device. It returns
-// Manager that uses this TUN device, along with an error channel which receives an error if the
-// admin subcommand stops running. Setup should be used in conjunction with Run.
+// SetupAndRun creates a network stack for VNet and runs it in the background. To do this, it also
+// needs to launch an admin subcommand in the background. It returns [ProcessManager] which controls
+// the lifecycle of both background tasks.
 //
-// adminCommandCtx is the context tied to the lifecycle of the admin subcommand. tsh typically
-// passes the same context as both ctx and adminCommandCtx, whereas Connect uses two different
-// contexts – one that is bound to an RPC that starts VNet and another that is bound to a separate
-// goroutine which runs Manager.
-func Setup(ctx context.Context, adminCommandCtx context.Context, appProvider AppProvider) (*Manager, <-chan error, error) {
+// The caller is expected to call Close on the process manager to close the network stack and clean
+// up any resources used by it.
+//
+// ctx is used to wait for setup steps that happen before SetupAndRun hands out the control to the
+// process manager. If ctx gets canceled during SetupAndRun, the process manager gets closed along
+// with its background tasks.
+func SetupAndRun(ctx context.Context, appProvider AppProvider) (*ProcessManager, error) {
 	ipv6Prefix, err := IPv6Prefix()
 	if err != nil {
-		return nil, nil, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
+	dnsIPv6 := Ipv6WithSuffix(ipv6Prefix, []byte{2})
+
+	pm := newProcessManager()
+	success := false
+	defer func() {
+		if !success {
+			// Closes the socket and background tasks.
+			pm.Close()
```

Comment:
If we make it so that `Close` does not `Wait` underneath, would it be fine to just call `Close` here and not `Wait` for any errors?

Reply:
In 99% of cases the errors from `Wait` would be `context.Canceled` anyway, which I suppose we don't want to propagate if `SetupAndRun` because, for example, a socket couldn't have been created.

Reply:
I guess technically those two background tasks might react differently to a context being canceled and one could just forward `context.Canceled`, while the other would run into another problem while closing. Which I suppose was the original intention behind capturing all errors from `errgroup` and not just the first one, to avoid a race. 😣

Reply:
But in this specific scenario it shouldn't matter (I think) because `SetupAndRun` has failed anyway and we just want to clean up any resources.

Reply:
with Close not returning any error now i think this is fine